crates/turborepo-cache/src/cache_archive/restore_symlink.rs

```diff
@@ -0,0 +1,150 @@
+use std::{backtrace::Backtrace, path::Path};
+
+use camino::{Utf8Path, Utf8PathBuf};
+use path_clean::clean;
+use turbopath::{AbsoluteSystemPath, AnchoredSystemPath, AnchoredSystemPathBuf, PathError};
+
+use crate::{
+    cache_archive::{restore::canonicalize_name, restore_directory::CachedDirTree},
+    CacheError,
+};
+
+pub fn restore_symlink(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    header: &tar::Header,
+) -> Result<AnchoredSystemPathBuf, CacheError> {
+    let processed_name = canonicalize_name(&header.path()?)?;
+
+    let processed_linkname = canonicalize_linkname(
+        anchor,
+        &processed_name,
+        &header.link_name()?.expect("has linkname"),
+    )?;
+
+    if processed_linkname.symlink_metadata().is_err() {
+        return Err(CacheError::LinkTargetDoesNotExist(
+            processed_linkname.to_string(),
+            Backtrace::capture(),
+        ));
+    }
+
+    actually_restore_symlink(dir_cache, anchor, &processed_name, header)?;
+
+    Ok(processed_name)
+}
+
+pub fn restore_symlink_allow_missing_target(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    header: &tar::Header,
+) -> Result<AnchoredSystemPathBuf, CacheError> {
+    let processed_name = canonicalize_name(&header.path()?)?;
+
+    actually_restore_symlink(dir_cache, anchor, &processed_name, header)?;
+
+    Ok(processed_name)
+}
+
+fn actually_restore_symlink<'a>(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    processed_name: &'a AnchoredSystemPath,
+    header: &tar::Header,
+) -> Result<&'a AnchoredSystemPath, CacheError> {
+    dir_cache.safe_mkdir_file(anchor, processed_name)?;
+
+    let symlink_from = anchor.resolve(processed_name);
+
+    _ = symlink_from.remove();
+
+    let link_name = header.link_name()?.expect("have linkname");
+    let symlink_to = link_name.to_str().ok_or_else(|| {
+        CacheError::PathError(
+            PathError::InvalidUnicode(link_name.to_string_lossy().to_string()),
+            Backtrace::capture(),
+        )
+    })?;
+
+    if Utf8Path::new(symlink_to).is_dir() {
+        symlink_from.symlink_to_file(symlink_to)?;
+    } else {
+        symlink_from.symlink_to_dir(symlink_to)?;
+    }
+
+    #[cfg(target_os = "macos")]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        let metadata = symlink_from.symlink_metadata()?;
+        let mut permissions = metadata.permissions();
+        if let Ok(mode) = header.mode() {
+            permissions.set_mode(mode);
+        }
+    }
+
+    Ok(processed_name)
+}
+
+// canonicalizeLinkname determines (lexically) what the resolved path on the
+// system will be when linkname is restored verbatim.
+pub fn canonicalize_linkname(
+    anchor: &AbsoluteSystemPath,
+    processed_name: &AnchoredSystemPathBuf,
+    linkname: &Path,
+) -> Result<Utf8PathBuf, CacheError> {
```

Comment:
I think we need to be careful with this function. We're introducing a failure mode that doesn't exist on the Go side. `canonicalizeLinkname` _always_ returns a `string`. Do we actually need the restriction that symlink targets (which may be garbage!) are valid utf8 strings? It looks like if we error here, we fail the whole restoration, rather than just deferring symlink creation?

Reply:
I think we should keep the requirement that all paths are UTF-8. If I remember correctly, we don't restore symlinks outside of the repo directory, so any symlink target is inside the repo. Having any path in the repo be valid UTF-8 is a reasonable rule IMO.